### ISSUE TITLE
FITS image writing fails for gzip file output with non-contiguous array

### ISSUE THREAD

**Issue by taldcroft** [astropy/astropy#8726]
When writing an image array that is not contiguous in memory (e.g. has been transposed) to a gzip file output, the in-memory order is used and the output image is corrupted.  The shape is still correct.  I spent a while getting down to this bug in my code because I could not imagine that `io.fits` was having such a basic problem...  And it doesn't make sense that the gzip at the end would make a difference, but maybe gzip here is not what I think?
```
img = np.arange(6).reshape(3, 2)
img = img.transpose()  # If we .copy() here then bug goes away
print(img.shape)
print(repr(img))
print()

for filename in ('test.fits', 'test.fits.gz'):
    hdu = fits.PrimaryHDU(img)
    hdu.writeto(filename, overwrite=True)
    with fits.open(filename) as hdus:
        img2 = hdus[0].data
    print(filename)
    print(img2.shape)
    print(repr(img2))
    print('Equal: ', np.all(img == img2))
    print()

(2, 3)
array([[0, 2, 4],
       [1, 3, 5]])

test.fits
(2, 3)
array([[0, 2, 4],
       [1, 3, 5]])
Equal:  True

test.fits.gz
(2, 3)
array([[0, 1, 2],
       [3, 4, 5]])
Equal:  False
```
#2150 is an old and somewhat-related issue.

**Comment by saimn** [comment]
This looks the same as the issue reported by @cdeil (https://github.com/astropy/astropy/issues/2150#issuecomment-375369793), but maybe not the same as the original issue in #2150 which was concerning compressed HDUs. 

And I think that this issue has been fixed by #8282: I get no error with @cdeil 's example and yours.

Which astropy version are you using?

**Comment by taldcroft** [comment]
Do'ooh, I was using an older version of astropy.  Sorry for the noise, closing.
